### ISSUE TITLE
Errors when cross compiling for x86_64 using CMake

### ISSUE THREAD

**Issue by crayroud-nviso** [xianyi/OpenBLAS#2245]
In another ticket (#2235) we started to discuss about the issue, but I can see that in the generated config.h file there is still less content than for a x86_64 host platform

I want to build OpenBLAS for an Intel NUC platform running Linux (Yocto). This is then a cross compile use case for x86_64. I have set the TARGET to CORE2. 

Here is the error that I have: "undefined symbol SLOCAL_BUFFER_SIZE"

Here is what I see missing in config.h compare  x86_64 host platform (for sure not all are needed):

```
#define SLOCAL_BUFFER_SIZE	24576
#define DLOCAL_BUFFER_SIZE	32768
#define CLOCAL_BUFFER_SIZE	12288
#define ZLOCAL_BUFFER_SIZE	8192

#define L1_CODE_SIZE 32768
#define L1_CODE_ASSOCIATIVE 8
#define L1_CODE_LINESIZE 64
#define L1_DATA_ASSOCIATIVE 8
#define L2_ASSOCIATIVE 8
#define ITB_SIZE 2097152
#define ITB_ASSOCIATIVE 0
#define ITB_ENTRIES 8
#define DTB_ASSOCIATIVE 4
#define HAVE_SSE4_1
#define HAVE_SSE4_2
#define HAVE_AVX
#define HAVE_AVX2
#define HAVE_FMA3
#define HAVE_CFLUSH
#define NUM_SHAREDCACHE 2
#define NUM_CORES 8
#define CORE_HASWELL
#define CHAR_CORENAME "HASWELL"
```

**Comment by martin-frbg** [comment]
The original ticket was automatically closed by my commit, but reopening it should have been possible. I'll do another PR for the xLOCAL_BUFFER_SIZE, the other parameters you listed should all be either not applicable to CORE2 (e.g. AVX) or not needed. (The cpuid code used by the getarch tool probably came from a different project originally - it outputs a lot more details than are actually used anywhere in OpenBLAS or the original GotoBLAS).

**Comment by martin-frbg** [comment]
Should work now with the latest commit and `-DTARGET=CORE2 -DARCH=x86_64` (though I only simulated the cross-compilation by adding `-DCMAKE_SYSTEM_NAME=Linux`). 
Another surprise discovery from this issue is that `cmake -DTARGET=...` was not working correctly even when not cross-compiling - it would still use `getarch` as normal, combining the desired target kernels with configuration data obtained from the cpu of the build host.

**Comment by crayroud-nviso** [comment]
I had to add following to prebuild.cmake, is it ok ?

```
set(CGEMM3M_UNROLL_M 4)
set(CGEMM3M_UNROLL_N 2)
set(ZGEMM3M_UNROLL_M 2)
set(ZGEMM3M_UNROLL_N 2)
```

**Comment by martin-frbg** [comment]
Strange, this should not be necessary (I think) - defaults should be copied from the respective SGEMM and DGEMM unroll values automatically (so should be 8,4,4,4 for CORE2) . My test build passed without them. (Possibly your experience comes from cmake not honoring the TARGET argument originally ?)

**Comment by crayroud-nviso** [comment]
It should not be copy from CGEMM and ZGEMM instead of S and D ?

I'm using TARGET=CORE2 argument and DYNAMIC_ARCH=OFF

**Comment by martin-frbg** [comment]
If you compare kernel/x86_64/KERNEL.CORE2 it is indeed using an 8x4 kernel for cgemm3m and 4x4 for zgemm3m. I'll repeat my build test to be sure, but with TARGET=CORE2 (and the patch from PR #2246) I believe these parameters are not required.

**Comment by martin-frbg** [comment]
Actually my pseudo-cross build manages to omit the GEMM3M functions (which are only built when X86 or X86_64 is defined - which probably does not happen due to the "cross" nature of the build).

**Comment by martin-frbg** [comment]
Not sure if my test setup is flawed - do you have to use `-DARCH=x86_64` or `-DX86_64=1` along with the `TARGET=CORE2` ? (I wonder if it would be useful to derive this from TARGET in system_check.cmake when CMAKE_CROSSCOMPILING is set and CMAKE_SYSTEM_PROCESSOR cannot be used to identify the architecture ?)
And another unrelated oddity - it seems to me that `-DBINARY=32` is currently not working correctly in our cmake setup, the `-m32` gets added to the fortran options but somehow never makes it to CCOMMON_OPT...

**Comment by crayroud-nviso** [comment]
No we are not using -DARCH=x86_64 or -DX86_64=1

**Comment by martin-frbg** [comment]
I believe this was fixed by #2248 (?)

**Comment by crayroud-nviso** [comment]
Yes
